xdsl/frontend/jaxpr/__init__.py

```diff
@@ -26,58 +47,105 @@ class IRGen:
 
+    symbol_table: ScopedSymbolTable | None = None
+    """
+    The symbol table maps a variable name to a value in the current scope.
+    Entering a function creates a new scope, and the function arguments are
+    added to the mapping. When the processing of a function is terminated, the
+    scope is destroyed and the mappings created in this scope are dropped."""
+
     def __init__(self):
         # We create an empty MLIR module and codegen functions one at a time and
         # add them to the module.
         self.module = ModuleOp([])
+        self.primitive_builder_jax = PrimitiveBuilder()
         self.builder = Builder.at_end(self.module.body.blocks[0])
 
+    def declare(self, var: str, value: SSAValue) -> bool:
+        """
+        Declare a variable in the current scope, return success if the variable
+        wasn't declared yet."""
+        assert self.symbol_table is not None
+        if var in self.symbol_table:
+            return False
+        self.symbol_table[var] = value
+        return True
+
+    def ir_gen_variable_expr(self, name: str) -> SSAValue:  # pyright: ignore[reportReturnType]
+        """
+        This is a reference to a variable in an expression. The variable is
+        expected to have been declared and so should have a value in the symbol
+        table, otherwise emit an error and return nullptr."""
+        assert self.symbol_table is not None
+        try:
+            variable = self.symbol_table[name]
+            return variable
+        except Exception as e:
+            self.error(f"error: unknown variable `{name}`", e)
+
+    def insert_return_op(self, return_vars: list[str]) -> None:
+        assert len(return_vars) == 1
+        with ImplicitBuilder(self.builder):
+            return_ssa = self.ir_gen_variable_expr(return_vars[0])
+            Return(return_ssa)
+
     def ir_gen_module(self, jaxpr: ClosedJaxpr) -> ModuleOp:
         """
         Public API: convert the jaxpr (source file) to an MLIR
         Module operation."""
         parent_builder = self.builder
+        self.symbol_table = ScopedSymbolTable()
 
-        assert len(jaxpr.jaxpr.invars) == 1
         assert len(jaxpr.jaxpr.outvars) == 1
+
         for arg in jaxpr.jaxpr.invars:
             assert arg.aval.dtype == "float32"
 
-        inputVars = jaxpr.jaxpr.invars[0]
-        outputVars = jaxpr.jaxpr.outvars[0]
+        inputVars = jaxpr.jaxpr.invars
+        outputVars = jaxpr.jaxpr.outvars
+
+        for i in range(1, len(inputVars)):
+            assert inputVars[i].aval.shape == inputVars[i - 1].aval.shape
 
         input_types = [
-            TensorType(f32, [inputVars.aval.size])
+            TensorType(f32, inputVars[0].aval.shape)
             for _ in range(len(jaxpr.jaxpr.invars))
         ]
+
         output_types = [
-            TensorType(f32, [outputVars.aval.size])
+            TensorType(f32, outputVars[0].aval.shape)
             for _ in range(len(jaxpr.jaxpr.outvars))
         ]
```

Comment:
Shouldn't these be the shapes for each invar, not always the first one?

Reply:
As of now Im asserting that all invars are with same shape (assert inputVars[i].aval.shape == inputVars[i - 1].aval.shape). and Im using 1st invars to get shape.

Reply:
If they're equivalent, then why not use each shape since it seems simpler and more future-proof?

Reply:
okk. Ill update